File: tensorflow.js/train.js
import * as tf from '@tensorflow/tfjs-node';
import * as fs from 'fs';
import now from 'performance-now';

const ALPHA = 0.005,
	EPOCHS = 100,
	BATCH_SIZE = 300,
	IMG_X = 28,
	IMG_Y = 28,
	HIDDEN_SIZE = 16,
	NUM_DIGITS = 10;

function max (arr) {
	let maxIdx = -1;
	let maxV = -Infinity;

	for (let i = 0; i < arr.length; i++) {
		if (arr[i] > maxV) {
			maxIdx = i;
			maxV = arr[i];
		}
	}

	return [ maxIdx, maxV ];
}


function getData () {
	let [ X, y ] = JSON.parse(String(fs.readFileSync('../train.json')));
	let [ X_test, y_test ] = JSON.parse(String(fs.readFileSync('../test.json')));

	y = y.map(v => {
		let arr = Array(NUM_DIGITS).fill(-.1);
		arr[v] = 1;
		return arr;
	});

	y_test = y_test.map(v => {
		let arr = Array(NUM_DIGITS).fill(-.1);
		arr[v] = 1;
		return arr;
	});

	y = tf.tensor(y).reshape([X.length, NUM_DIGITS]);
	X = tf.tensor(X, [X.length, IMG_X, IMG_Y]);

	y_test = tf.tensor(y_test).reshape([X_test.length, NUM_DIGITS]);
	X_test = tf.tensor(X_test, [X_test.length, IMG_X, IMG_Y]);

	return [ X, y, X_test, y_test ];
}

async function test (model, X_test, y_test) {
	const n = Math.min(1000, X_test.shape[0]);

	y_test = await y_test.array();
	X_test = await X_test.array();

	let totalError = 0;
	let incorrects = 0;

	for (let i = 0; i < n; i++) {
		const X = tf.tensor(X_test[i]).reshape([1, 28, 28]);
		let y = (await model.predict(X).array())[0];
		for (let j = 0; j < NUM_DIGITS; j++) {
			totalError += (y[j] - y_test[i][j])**2;
		}

		if (max(y)[0] !== max(y_test[i])[0]) {
			incorrects++;
		}
	}

	console.log(`${incorrects}/${n} incorrect`);
	console.log(`${totalError/n} avg error`);
}

async function main () {
	const [X, y, X_test, y_test] = getData();
	const model = tf.sequential();

	model.add(
		tf.layers.dense({
			inputShape: [IMG_X, IMG_Y],
			units: HIDDEN_SIZE,
			activation: "tanh",
		})
	);
	model.add(
		tf.layers.dense({
			units: HIDDEN_SIZE,
			activation: "tanh",
		})
	);
	model.add(tf.layers.flatten());
	model.add(
		tf.layers.dense({
			units: NUM_DIGITS,
		})
	);

	model.compile({
		optimizer: tf.train.sgd(ALPHA),
		loss: "meanSquaredError",
	});

	let losses = [];
	let times = [];

	let start = now();

	await model.fit(X, y, {
		epochs: EPOCHS,
		callbacks: {
			onEpochEnd: async (epoch, logs) => {
				losses.push(logs.loss);
				times.push(now() - start);
				start = now();
				if (epoch % 100 === 0 || epoch === EPOCHS) {
					console.log(`Finished epoch ${epoch+1} / ${EPOCHS}: ${logs.loss}`);
				}
			},
		},
		batchSize: BATCH_SIZE,
		verbose: false,
		shuffle: true,
		validationData: [X_test, y_test]
	});

	console.log((times.reduce((a, b) => a + b, 0) / times.length).toFixed(5), 'ms on average');

	await test(model, X_test, y_test);

	await model.save('file:///home/joseph/dev/mnist/tensorflow.js/model');

	console.log('Saved model');
}

main();

File: tensorflow/train.py
import tensorflow as tf
import tensorflow_datasets as tfds
import tensorflowjs as tfjs

ALPHA = 0.001
EPOCHS = 2


def test(model, X_test, y_test):
    n = 100

    totalError = 0
    incorrects = 0

    for i in range(n):
        X = tf.constant(X_test[i]).set_shape([1, 28, 28])
        y = list(model.predict(X))[0]
        for j in range(10):
            totalError += (y[j] - y_test[i][j]) ** 2

        if max(y)[0] != max(y_test[i])[0]:
            incorrects += 1

    print(f'{incorrects}/{n} incorrect')
    print(f'{totalError / n} avg error')


def main():
    (ds_train, ds_test), ds_info = tfds.load(
        'mnist',
        split=['train', 'test'],
        shuffle_files=True,
        as_supervised=True,
        with_info=True,
    )

    def normalize_img(image, label):
        return tf.cast(image, tf.float32) / 255., label

    ds_train = ds_train.map(
        normalize_img, num_parallel_calls=tf.data.AUTOTUNE)
    ds_train = ds_train.cache()
    ds_train = ds_train.shuffle(ds_info.splits['train'].num_examples)
    ds_train = ds_train.batch(128)
    ds_train = ds_train.prefetch(tf.data.AUTOTUNE)

    ds_test = ds_test.map(
        normalize_img, num_parallel_calls=tf.data.AUTOTUNE)
    ds_test = ds_test.batch(128)
    ds_test = ds_test.cache()
    ds_test = ds_test.prefetch(tf.data.AUTOTUNE)

    model = tf.keras.models.Sequential([
        tf.keras.layers.Flatten(input_shape=(28, 28)),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dense(10)
    ])

    model.compile(
        optimizer=tf.keras.optimizers.Adam(ALPHA),
        loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
        metrics=[tf.keras.metrics.SparseCategoricalAccuracy()],
    )

    model.fit(
        ds_train,
        epochs=EPOCHS,
        validation_data=ds_test,
        verbose=False
    )

    model.save('file:///home/joseph/dev/mnist/tensorflow/py-model')
    tfjs.converters.save_keras_model(model, 'model')


if __name__ == '__main__':
    main()
